docs/architecture/psa-migration/psa-legacy-bridges.md

```diff
@@ -0,0 +1,321 @@
+
+When creating a PSA key, apart from the key material, the key is determined by attributes, which fall under three categories:
+
+* Type and size. These are directly related to the key material and can be deduced from it if the key material is in a structured format, which is the case with an `mbedtls_pk_context` input.
+* Policy. This includes the chosen algorithm, which as discussed above cannot be fully deduced from the `mbedtls_pk_context` object. Just choosing one algorithm is problematic because it doesn't allow implementation-specific extensions, such as Mbed TLS's enrollment algorithm. The intended usage flags cannot be deduced from the PK context either, but the conversion function could sensibly just enable all the relevant usage flags. Users who want a more restrictive usage can call `psa_copy_key` and `psa_destroy_key` to obtain a PSA key object with a more restrictive usage.
+* Persistence and location. This is completely orthogonal to the information from the `mbedtls_pk_context` object. It is convenient, but not necessary, for the conversion function to allow customizing these aspects. If it doesn't, users can call the conversion function and then call `psa_copy_key` and `psa_destroy_key` to move the key to its desired location.
+
+To allow the full flexibility around policies, and make the creation of a persistent key more convenient, the conversion function shall take a `const psa_key_attributes_t *` input, like all other functions that create a PSA key. In addition, there shall be a helper function to populate a `psa_key_attributes_t` with a sensible default. This lets the caller choose a more flexible, or just different usage policy, unlike the default-then-copy approach which only allows restricting the policy.
+
+This is close to the existing function `mbedtls_pk_wrap_as_opaque`, but does not bake in the implementation-specific consideration that a PSA key has exactly two algorithms, and also allows the caller to benefit from default for the policy in more cases.
+
+[ACTION] Implement `mbedtls_pk_get_psa_attributes` and `mbedtls_pk_import_into_psa` as described below. These functions are available whenever `MBEDTLS_PK_C` and `MBEDTLS_PSA_CRYPTO_CLIENT` are both defined. Deprecate `mbedtls_pk_wrap_as_opaque`.
```

Comment:
Why `USE_PSA_CRYPTO`? I don't see why it would help when implementing the function. While configurations without `USE_PSA_CRYPTO` are lower-priority, they do matter: not everyone controls how the library is built. So I think we shouldn't require `USE_PSA_CRYPTO` without a good reason.

Reply:
Mainly for the sake of uniformity, as I said: currently all uses of PSA in PK are gated by `USE_PSA_CRYPTO`. But perhaps that's not a good enough reason.

Now that I think about it, currently `USE_PSA_CRYPTO` does two things:
1. Change the behaviour of existing functions so that they call PSA internally - this has to be explicitly opt-in because it means the user needs to call `psa_crypto_init()` before calling these functions.
2. Make new functions available (such as `pk_setup_opaque()`). This does not really need a specific option, and we could make those functions available as soon as PSA enabled in the build.

If the new functions proposed by this document are available as soon as PSA is enabled in the build, then perhaps we should change the guards around `pk_setup_opaque()` as well? From a user's perspective, it would make little sense for `mbedtls_pk_import_into_psa()` to be guarded by `MBEDTLS_PSA_CRYPTO_CLIENT` but `mbedtls_pk_setup_opaque()` by `MBEDTLS_USE_PSA_CRYPTO`.

Reply:
> Make new functions available (such as pk_setup_opaque()

I was never under the impression that this was part of the intended scope of `MBEDTLS_USE_PSA_CRYPTO`. I would only expect that option to enable new functions if those functions are directly related to using PSA keys in X.509 and TLS, for example an API to directly pass a PSA key identifier to a TLS function.

`mbedtls_pk_setup_opaque` looks like a separate feature to choose at runtime to make a pk object that calls down to PSA. This doesn't seem related to `MBEDTLS_USE_PSA_CRYPTO`, which is about compile-time choices. So I don't see a reason why it wouldn't be available whenever `MBEDTLS_PK_C` and `MBEDTLS_PSA_CRYPTO_CLIENT` are both defined.

I guess I should add an action to provide `mbedtls_pk_setup_opaque` whenever `MBEDTLS_PSA_CRYPTO_CLIENT` is enabled? This doesn't feel very important, but it's a weird gap.

Reply:
So, out of curiosity I had a look, and currently the following public functions depends on `USE_PSA_CRYPTO`:
- `mbedtls_cipher_setup_psa()` (deprecated)
- `mbedtls_pk_setup_opaque()`
- `mbedtls_pk_can_do_ext()`
- `mbedtls_pk_wrap_as_opaque()` (documented as temporary, will be replaced soon)
- `mbedtls_ssl_conf_psk_opaque()` and `mbedtls_ssl_set_hs_psk_opaque()`

> I was never under the impression that this was part of the intended scope of MBEDTLS_USE_PSA_CRYPTO. I would only expect that option to enable new functions if those functions are directly related to using PSA keys in X.509 and TLS, for example an API to directly pass a PSA key identifier to a TLS function.

Note that `mbedtls_pk_setup_opaque()` was introduced as a way to pass a PSA key to a TLS function (indirectly). But I agree that in the above list, `USE_PSA` should only control the last item. PK functions could be changed to `MBEDTLS_PSA_CRYPTO_CLIENT`, and I wouldn't both changing the cipher one as it's already deprecated anyway.

> I guess I should add an action to provide `mbedtls_pk_setup_opaque` whenever `MBEDTLS_PSA_CRYPTO_CLIENT` is enabled? This doesn't feel very important, but it's a weird gap.

Agreed: it's not important, but it would be nice for things to be consistent.

Reply:
Note: if we change guards around `mbedtls_pk_setup_opaque()` then we'll want to update [the documentation of `USE_PSA`](https://github.com/Mbed-TLS/mbedtls/blob/development/docs/use-psa-crypto.md) as well.